src/py/flwr/server/serverapp/app.py

```diff
@@ -121,21 +142,94 @@ def _try_obtain_certificates(
     return root_certificates
 
 
-def run_serverapp(  # pylint: disable=R0914
+def run_serverapp(  # pylint: disable=R0914, disable=W0212
     superlink: str,
     run_id: Optional[int] = None,
     flwr_dir_: Optional[str] = None,
     certificates: Optional[bytes] = None,
 ) -> None:
     """Run Flower ServerApp process."""
-    _ = GrpcDriver(
+    driver = GrpcDriver(
         run_id=run_id if run_id else 0,
         driver_service_address=superlink,
         root_certificates=certificates,
     )
 
-    log(INFO, "%s", flwr_dir_)
+    # Resolve directory where FABs are installed
+    flwr_dir = get_flwr_dir(flwr_dir_)
 
-    # Then, GetServerInputs
+    only_once = run_id is not None
 
-    # Then, run ServerApp
+    while True:
+
+        try:
+            # Pull ServerAppInputs from LinkState
+            req = (
+                PullServerAppInputsRequest(run_id=run_id)
+                if run_id
+                else PullServerAppInputsRequest()
+            )
```

Comment:
Isn't the first step to pull a `run_id`?

Reply:
not in the current logic. `PullServerAppInputsRequest` returns any pending run if no `run_id` is set